abci/client/client.go

```diff
@@ -27,35 +28,35 @@ type Client interface {
 	SetResponseCallback(Callback)
 	Error() error
 
-	FlushAsync() *ReqRes
-	EchoAsync(msg string) *ReqRes
-	InfoAsync(types.RequestInfo) *ReqRes
-	DeliverTxAsync(types.RequestDeliverTx) *ReqRes
-	CheckTxAsync(types.RequestCheckTx) *ReqRes
-	QueryAsync(types.RequestQuery) *ReqRes
-	CommitAsync() *ReqRes
-	InitChainAsync(types.RequestInitChain) *ReqRes
-	BeginBlockAsync(types.RequestBeginBlock) *ReqRes
-	EndBlockAsync(types.RequestEndBlock) *ReqRes
-	ListSnapshotsAsync(types.RequestListSnapshots) *ReqRes
-	OfferSnapshotAsync(types.RequestOfferSnapshot) *ReqRes
-	LoadSnapshotChunkAsync(types.RequestLoadSnapshotChunk) *ReqRes
-	ApplySnapshotChunkAsync(types.RequestApplySnapshotChunk) *ReqRes
-
-	FlushSync() error
-	EchoSync(msg string) (*types.ResponseEcho, error)
-	InfoSync(types.RequestInfo) (*types.ResponseInfo, error)
-	DeliverTxSync(types.RequestDeliverTx) (*types.ResponseDeliverTx, error)
-	CheckTxSync(types.RequestCheckTx) (*types.ResponseCheckTx, error)
-	QuerySync(types.RequestQuery) (*types.ResponseQuery, error)
-	CommitSync() (*types.ResponseCommit, error)
-	InitChainSync(types.RequestInitChain) (*types.ResponseInitChain, error)
-	BeginBlockSync(types.RequestBeginBlock) (*types.ResponseBeginBlock, error)
-	EndBlockSync(types.RequestEndBlock) (*types.ResponseEndBlock, error)
-	ListSnapshotsSync(types.RequestListSnapshots) (*types.ResponseListSnapshots, error)
-	OfferSnapshotSync(types.RequestOfferSnapshot) (*types.ResponseOfferSnapshot, error)
-	LoadSnapshotChunkSync(types.RequestLoadSnapshotChunk) (*types.ResponseLoadSnapshotChunk, error)
-	ApplySnapshotChunkSync(types.RequestApplySnapshotChunk) (*types.ResponseApplySnapshotChunk, error)
+	FlushAsync() (*ReqRes, error)
```

Comment:
Maybe asyncs should take a context too?

Reply:
Thought about it too. But what for? There's nothing to break from. We could add a context and allow the caller to control whatever it wants to block (context.Background), wait for 5s (context.WithTimeout), or return immediately (nil context), but I felt like it would complicate things unnecessarily.

Reply:
With async, there's going to be a bunch of requests queued up somewhere, and under high contention it might be nice to have a mechanism to abort queued requests to reduce contention. For example, if an RPC client sends a bunch of requests, and then times out or goes away, there's no need to process the queued requests.

Reply:
good point! still I believe the simpler scheme is better: if there's space for request, we process it. if not, we return an error and "ask" a caller to repeat if needed.

Reply:
> For example, if an RPC client sends a bunch of requests, and then times out or goes away, there's no need to process the queued requests.

just to clarify what you're suggesting here is to store `context`, which will allow us to identify the client, along with request in the queue. Once the context is done, we'll need to find the request in the queue and remove it. Not sure we want to listen for all those contexts.

or we can pass a client ID in the context and create an additional callback (`DropAsyncRequests`), which will be called when client is stopped. Not sure we need to do this here though.

Reply:
so if we're breaking the API here, we should add `context.Context` to Async methods too for possible future optimizations, right?